### PR TITLE
Uniffi push component

### LINKED ISSUE #4424: Implement uniffi'ing the push component
This includes following the instructions here https://mozilla.github.io/application-services/book/howtos/converting-a-component-to-uniffi.html 

For this to be done, we should completely eliminate the protocol buffers and the handwritten FFI code.

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/SYNC-2763)
┆Story Points: 5
┆Epic: Uniffi Push Component
┆Fix Versions: Release 94
┆Sprint End Date: 2021-09-17

### PR DESCRIPTION
**Pull request by tarikeshaq**
Fixes #4424 

The commits in order:
- [x] adds a valid UDL
- [x] Rust compiles with scaffolding included
- [x] Removes ffi crate
- [x] Removes protobuf types and replaces them with rust types
- [x] Adds kotlin bindings and cleans up the API to allow android to consume this 

This is ready for review 🚀 Before this merges though, a couple of TODOS:
- [x] Create changelog entry describing the breaking changes (added: https://github.com/mozilla/application-services/pull/4431/commits/d6bf2e4186a789a9255ea70678bbafccd921b304)
- [x] Create Android components PR to fix breaking changes, get review there and make appropriate changes here (PR: https://github.com/mozilla-mobile/android-components/pull/10953)

### CODE REVIEW

File: components/push/src/push.udl
Comment (tarikeshaq) [reply components/push/src/push.udl:9]: also a reminder to myself about the breaking change here, the `bridgeType` used to be a `BridgeType` object, and our bindings used to `toString` it, now A-C will have to do that

Comment (tarikeshaq) [reply components/push/src/push.udl]: ended up getting rid of the breaking change 😄

File: components/push/android/src/test/java/mozilla/appservices/push/PushTest.kt
Comment (tarikeshaq) [inline components/push/android/src/test/java/mozilla/appservices/push/PushTest.kt:141]: The changes in this file are the extent of the breaking changes...

- `A-C` will have to turn the `List<Byte>` into a `ByteArray` (what `decrypt` used to return) if they'd like using `toByteArray`. It was either this, or have a small wrapper to do the conversion, I opted for the "less android code in a-s" approach
- `channelID` -> `channelId`: this is trivial but interesting, I tried setting it to `channel_ID`, but the rust compiler gave me warnings to turn it into `channel_id` to make it snake case. So either we make the breaking change, or we can silence the rust compiler... I almost opted for the latter, but wanted to get a second opinion

File: Cargo.toml
Comment (mhammond) [inline Cargo.toml:12]: 🚀

File: components/push/src/lib.rs
Comment (mhammond) [inline components/push/src/lib.rs:394]: Is this related to unsigned types being experimental in Kotlin? If so, I think that's no longer the case.  (IOW, if it's possible to now use u8 here I think we should do so, but I've no idea how practical that is, so feel free to ignore!)

**Review by mhammond — APPROVED** [review]
Great work Tarik - very thorough and well thought out! Please do complete those other todos before merging though - I'm worried about overloading the android-components people, so think we should have a PR up before merging.

But great work, thanks!

File: components/push/src/lib.rs
Comment (tarikeshaq) [reply components/push/src/lib.rs:394]: Uhh yes, I used this because Android components expects a `ByteArray`, which is just one `.toByteArray()` away from a `List<Byte>`, but `List<UByte>` can't convert directly. So A-C will have to convert the `List<UByte>` to `ByteArray` through a `UByteArray`... which isn't _that_ bad, but it would be the first time they interact with unsigned types over there 😅 I guess it's just a matter of do we want to do the conversion for them, or have them do the conversion themselves

**Comment by tarikeshaq** [comment]
Thanks Mark!! I created a pr on A-C to get feedback https://github.com/mozilla-mobile/android-components/pull/10953

Correct me if I'm wrong, but from yesterday's discussion, we should cut a release before this merges, resolve the _other_ breaking changes, then cut **another** release for this right?

**Comment by tarikeshaq** [comment]
This should be in a mergeable state! I'm planning to merge it sometime early next week and cut another major release. Once that's done I'll update https://github.com/mozilla-mobile/android-components/pull/10953 to have the latest release. (cc: @mhammond and @jonalmeida, no action items, just letting you know my plan here in case you have any concerns or thoughts 😄)

**Pull request merged**
